ProcessLib/RichardsComponentTransport/RichardsComponentTransportFEM.h

```diff
@@ -0,0 +1,444 @@
+
+            pos.setIntegrationPoint(ip);
+
+            auto const& K =
+                _process_data.porous_media_properties.getIntrinsicPermeability(
+                    t, pos);
+            auto const mu = _process_data.fluid_properties->getValue(
+                MaterialLib::Fluid::FluidPropertyType::Viscosity, vars);
+
+            double C_int_pt = 0.0;
+            double p_int_pt = 0.0;
+            NumLib::shapeFunctionInterpolate(local_x, N, C_int_pt, p_int_pt);
+
+            // saturation
+            double const pc_int_pt = -p_int_pt;
+            double const Sw =
+                (pc_int_pt > 0)
+                    ? _process_data.porous_media_properties
+                          .getCapillaryPressureSaturationModel(t, pos)
+                          .getSaturation(pc_int_pt)
+                    : 1.0;
+
+            auto const& k_rel = _process_data.porous_media_properties
+                                    .getRelativePermeability(t, pos)
+                                    .getValue(Sw);
+
+            cache_mat.col(ip).noalias() = -dNdx * p_nodal_values;
+            if (_process_data.has_gravity)
+            {
+                vars[static_cast<int>(
+                    MaterialLib::Fluid::PropertyVariableType::C)] = C_int_pt;
+                vars[static_cast<int>(
+                    MaterialLib::Fluid::PropertyVariableType::p)] = p_int_pt;
+
+                auto const rho_w = _process_data.fluid_properties->getValue(
+                    MaterialLib::Fluid::FluidPropertyType::Density, vars);
+                auto const b = _process_data.specific_body_force;
+                // here it is assumed that the vector b is directed 'downwards'
+                cache_mat.col(ip).noalias() += rho_w * b;
+            }
+            cache_mat.col(ip).noalias() = k_rel/mu * (K * cache_mat.col(ip));
+        }
+        return cache;
+    }
+
+    Eigen::Map<const Eigen::RowVectorXd> getShapeMatrix(
+        const unsigned integration_point) const override
+    {
+        auto const& N = _ip_data[integration_point].N;
+
+        // assumes N is stored contiguously in memory
+        return Eigen::Map<const Eigen::RowVectorXd>(N.data(), N.size());
+    }
+
+    std::vector<double> const& getIntPtSaturation(
+        const double t,
+        GlobalVector const& current_solution,
+        NumLib::LocalToGlobalIndexMap const& dof_table,
+        std::vector<double>& cache) const override
+    {
+        SpatialPosition pos;
+        pos.setElementID(_element_id);
+
+        unsigned const n_integration_points =
+            _integration_method.getNumberOfPoints();
+
+        auto const indices = NumLib::getIndices(_element_id, dof_table);
+        assert(!indices.empty());
+        auto const local_x = current_solution.get(indices);
+
+        cache.clear();
+        auto cache_vec = MathLib::createZeroedMatrix<
+            Eigen::Matrix<double, 1, Eigen::Dynamic>>(
+            cache, 1, n_integration_points);
```

Comment:
Not sure if the mapping is needed, because the cache is accessed only once by index (of integration point). But you might want to keep it if this style already became idiomatic.
The access below would be `cache[ip] = Sw`.